doc/source/reference/arrays.classes.rst

```diff
@@ -39,77 +39,88 @@ Special attributes and methods
-      than multiplication.
-
-      In such a case, when computing a binary operation such as
-      ``ndarray_obj * your_obj``, your ``__numpy_ufunc__`` method
-      *will not* be called.  Instead, the execution passes on to your
-      right-hand ``__rmul__`` operation, as per standard Python
-      operator override rules.
-
-      Similar special case applies to *in-place operations*: If you
-      define ``__rmul__``, then ``ndarray_obj *= your_obj`` *will not*
-      call your ``__numpy_ufunc__`` implementation. Instead, the
-      default Python behavior ``ndarray_obj = ndarray_obj * your_obj``
-      occurs.
-
-      Note that the above discussion applies only to Python's builtin
-      binary operation mechanism. ``np.multiply(ndarray_obj,
-      your_obj)`` always calls only your ``__numpy_ufunc__``, as
-      expected.
-
-.. method:: class.__array_finalize__(obj)
+   .. note:: In addition to ufuncs, :func:`__array_ufunc__` also
+      overrides the behavior of :func:`numpy.dot` and :func:`numpy.matmul`.
+      This even though these are not ufuncs, but they can be thought of as
+      :ref:`generalized universal functions<c-api.generalized-ufuncs>`
+      (which are overridden). We intend to extend this behaviour to other
+      relevant functions.
+
+   Like with other methods in python, such as ``__hash__`` and
+   ``__iter__``, it is possible to indicate that your class does *not*
+   support ufuncs by setting ``__array_ufunc__ = None``. With this,
+   inside ufuncs, your class will be treated as if it returned
+   :obj:`NotImplemented` (which will lead to an :exc:`TypeError`
+   unless another class also provides a :func:`__array_ufunc__` method
+   which knows what to do with your class).
+
+   The presence of :func:`__array_ufunc__` also influences how binary
+   and comparison operators are dealt with, such as ``__add__``,
+   ``__gt__``, etc.  If it is not :obj:`None`, the assumption is that
+   your code can handle such operations via the ufunc mechanism, and
+   hence forward methods on :class:`ndarray` will call the ufuncs
+   unconditionally (i.e., even if your class has defined reverse
+   methods such as ``__radd__``, ``__le__``, etc.). If
+   ``__array_ufunc__ = None``, however, forward methods on
+   :class:`ndarray` will unconditionally return :obj:`NotImplemented`,
+   so that your reverse methods will get called.
+
+   .. note:: If you subclass :class:`ndarray`:
+
+      - We strongly recommend that you avoid confusion by neither setting
+        :func:`__array_ufunc__` to :obj:`None`, which makes no sense for
+        an array subclass, nor by defining it and also defining reverse
+        methods, which methods will be called by ``CPython`` in
+        preference over the :class:`ndarray` forward methods.
+      - :class:`ndarray` defines its own :func:`__array_ufunc__`, which
+        corresponds to ``getattr(ufunc, method)(*inputs, **kwargs)``. Hence,
```

Comment:
This seems really wrong? `ndarray.__array_ufunc__` should return `NotImplemented` if any of its arguments are unrecognized (which I guess means "define their own `__array_ufunc__` or something? it's pretty subtle but it's definitely not the same as just calling the ufunc again).